src/trans-word-level/property.cpp

```diff
@@ -61,3 +62,128 @@ void property(
     prop_handles.push_back(std::move(handle));
   }
 }
+
+/*******************************************************************\
+
+Function: states_equal
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+exprt states_equal(
+  std::size_t i,
+  std::size_t k,
+  const std::vector<symbol_exprt> &state_vars)
+{
+  exprt::operandst conjuncts;
+  conjuncts.reserve(state_vars.size());
+
+  for(auto &var : state_vars)
+  {
+    auto i_var = timeframe_symbol(i, var);
+    auto k_var = timeframe_symbol(k, var);
+    conjuncts.push_back(equal_exprt(i_var, k_var));
+  }
+
+  return conjunction(std::move(conjuncts));
+}
+
+/*******************************************************************\
+
+Function: lasso_symbol
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+symbol_exprt lasso_symbol(std::size_t timeframe)
+{
+  irep_idt lasso_identifier = "lasso::"+std::to_string(timeframe);
+  return symbol_exprt(lasso_identifier, bool_typet());
+}
+
+/*******************************************************************\
+
+Function: lasso_constraints
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void lasso_constraints(
+  decision_proceduret &solver,
+  std::size_t no_timeframes,
+  const namespacet &ns,
+  const irep_idt &module_identifier)
+{
+  // gather the state variables
+  std::vector<symbol_exprt> state_vars;
+  const symbol_tablet &symbol_table=ns.get_symbol_table();
+
+  auto lower = symbol_table.symbol_module_map.lower_bound(module_identifier);
+  auto upper = symbol_table.symbol_module_map.upper_bound(module_identifier);
+
+  for(auto it=lower;
+      it!=upper;
+      it++)
+  {
+    const symbolt &symbol=ns.lookup(it->second);
+
+    if(symbol.is_state_var)
+      state_vars.push_back(symbol.symbol_expr());
+  }
+
+  for(std::size_t i = 1; i < no_timeframes; i++)
+  {
+    // Is there a loop back to any time frame k with k<i?
+    exprt::operandst disjuncts;
+    disjuncts.reserve(i - 1);
+
+    for(std::size_t k = 0; k < i; k++)
+      disjuncts.push_back(states_equal(k, i, state_vars));
+
+    auto lasso_symbol = ::lasso_symbol(i);
+    solver.set_to_true(equal_exprt(lasso_symbol, disjunction(disjuncts)));
+  }
+}
+
+/*******************************************************************\
+
+Function: requires_lasso_constraints
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+bool requires_lasso_constraints(const exprt &expr)
+{
+  if(
+    expr.id() == ID_sva_until || expr.id() == ID_sva_s_until ||
+    expr.id() == ID_sva_eventually || expr.id() == ID_sva_s_eventually)
+  {
+    return true;
+  }
+
+  for(auto &op : expr.operands())
+    if(requires_lasso_constraints(op))
+      return true;
+
+  return false;
+}
```

Comment:
I'm not really sure what the size of `expr` could end up being, but I think this would be a case where a depth iterator can easily be used to avoid the recursion.

Reply:
done